vortex-expr/src/lib.rs

```diff
@@ -1,27 +1,71 @@
 #![feature(iter_intersperse)]
 
+use std::any::Any;
+use std::collections::HashSet;
+use std::fmt::Debug;
 use std::sync::Arc;
 
+mod binary;
+mod column;
 pub mod datafusion;
-mod expr;
+mod identity;
+mod literal;
 mod operators;
+mod select;
 
-pub use expr::*;
+pub use binary::*;
+pub use column::*;
+pub use identity::*;
+pub use literal::*;
 pub use operators::*;
+pub use select::*;
+use vortex::Array;
+use vortex_dtype::field::Field;
+use vortex_error::{VortexExpect, VortexResult};
 
+/// [`VortexExpr`] represents logical operation on [`Array`]s
+pub trait VortexExpr: Debug + Send + Sync + PartialEq<dyn Any> {
+    fn as_any(&self) -> &dyn Any;
+
+    fn evaluate(&self, batch: &Array) -> VortexResult<Array>;
+
+    fn references(&self) -> HashSet<Field>;
```

Comment:
Add a `collect_references(&self, mut HashSet<Field>)` and call that from this function. Otherwise lots of allocations